src/main.rs

```diff
@@ -1,98 +1,173 @@
 
-struct XorHasher {
+struct EvanHash {
     state: [u8; HASH_SIZE],
-    block_count: usize,
+    block: [u8; BLOCK_SIZE],
+    length: usize,
+    data: Vec<u8>,
+    debug: bool,
 }
 
-impl XorHasher {
-    fn new() -> Self {
-        XorHasher {
-            state: [0; HASH_SIZE],
-            block_count: 0,
-        }
+impl EvanHash {
+    fn new(data: &[u8], debug: bool) -> Self {
+      let hasher = EvanHash {
+        state: INITIAL_STATE,
+        block: [0u8; BLOCK_SIZE],
+        length: 0,
+        data: Vec::from(data),
+        debug: debug,
+      };
+      hasher
     }
 
-    fn update(&mut self, data: &[u8]) {
-        let mut offset = 0;
-
-        while offset < data.len() {
-            let mut block = [0u8; BLOCK_SIZE];
-            let remaining = data.len() - offset;
-            let block_size = remaining.min(BLOCK_SIZE);
+    fn update(&mut self) {
+        self.block = [0u8; BLOCK_SIZE];
+        let remaining = self.data.len() - self.length;
+        let end = self.length + remaining.min(BLOCK_SIZE);
 
-            block[..block_size].copy_from_slice(&data[offset..offset + block_size]);
-            self.process_block(&block);
+        // fill the block from input data
+        self.block[..remaining.min(BLOCK_SIZE)].copy_from_slice(&self.data[self.length..end]);
 
-            offset += block_size;
-            self.block_count += 1;
+        // if this is the last block append data.len()
+        if remaining < BLOCK_SIZE {
+            self.block[remaining] = self.data.len() as u8;
         }
+
+        debug_print(self.debug, format!("block{:<7} {:?}", format!("{}:", self.length / BLOCK_SIZE), self.block));
+        self.process_block();
+
+        self.length += BLOCK_SIZE;
     }
 
     fn finalize(self) -> [u8; HASH_SIZE] {
         self.state
     }
 
-    fn process_block(&mut self, block: &[u8; BLOCK_SIZE]) {
-        for i in 0..HASH_SIZE {
-            // since we have HASH_SIZE == BLOCK_SIZE this is easy
-            self.state[i] ^= block[i];
-        }
+    fn process_block(&mut self) {
+      let mut temp_states: Vec<[u8; BLOCK_SIZE]> = Vec::new();
+      temp_states.push(self.state);
+  
+      for i in 0..13 {
+          let mut temp_block = self.block;
+          let block_as_u64: u64 = u64::from_be_bytes(temp_block.try_into().unwrap()); // Convert block to u64 for bitwise operations
+  
+          let modified_block_as_u64 = match i {
+              0 => !block_as_u64, // invert bits
+              1 => block_as_u64.rotate_right(17),
+              2 => block_as_u64.rotate_right(23),
+              3 => block_as_u64.rotate_right(32),
+              4 => block_as_u64.reverse_bits(),
+              5 => block_as_u64 ^ 0xAAAAAAAAAAAAAAAA, // Toggle every other bit starting with the first bit
+              6 => block_as_u64.rotate_right(43),
+              7 => block_as_u64 ^ 0x5555555555555555, // Toggle every other bit skipping the first bit
```

Comment:
5 and 7 actually cancel each other out, perhaps I should do something like combine invert and shift or combine reverse and shift